cpp/src/copying/shift.cu

```diff
@@ -0,0 +1,162 @@
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+namespace cudf {
+namespace experimental {
+namespace {
+
+template<typename T>
+struct value_functor {
+    column_device_view const input;
+    size_type size;
+    size_type offset;
+    T const* fill;
+
+    T __device__ operator()(size_type idx) {
+        auto src_idx = idx - offset;
+        return src_idx < 0 || src_idx >= size
+            ? *fill
+            : input.element<T>(src_idx);
+    }
+};
+
+struct validity_functor {
+    column_device_view const input;
+    size_type size;
+    size_type offset;
+    bool const* fill;
+
+    bool __device__ operator()(size_type idx) {
+        auto src_idx = idx - offset;
+        return src_idx < 0 || src_idx >= size
+            ? *fill
+            : input.is_valid(src_idx);
+    }
+};
+
+struct functor {
+
+    template<typename T, typename... Args>
+    std::enable_if_t<not cudf::is_fixed_width<T>(), std::unique_ptr<column>>
+    operator()(Args&&... args)
+    {
+        throw cudf::logic_error("shift does not support non-fixed-width types.");
+    }
+
+    template<typename T>
+    std::enable_if_t<cudf::is_fixed_width<T>(), std::unique_ptr<column>>
+    operator()(column_view const& input,
+               size_type offset,
+               scalar const& fill_value,
+               rmm::mr::device_memory_resource *mr,
+               cudaStream_t stream)
+    {
+        using ScalarType = cudf::experimental::scalar_type_t<T>;
+        auto& scalar = static_cast<ScalarType const&>(fill_value);
+
+        auto device_input = column_device_view::create(input);
+        auto output = allocate_like(input, mask_allocation_policy::NEVER);
+        auto device_output = mutable_column_device_view::create(*output);
+
+        auto index_begin = thrust::make_counting_iterator<size_type>(0);
+        auto index_end = thrust::make_counting_iterator<size_type>(input.size());
+
+        auto func_value = value_functor<T>{*device_input,
+                                           input.size(),
+                                           offset,
+                                           scalar.data() };
+
+        if (scalar.is_valid() && not input.nullable())
+        {
+            thrust::transform(rmm::exec_policy(stream)->on(stream),
+                              index_begin,
+                              index_end,
+                              device_output->data<T>(),
+                              func_value);
+
+            return output;
+        }
+
+
+        auto func_validity = validity_functor{*device_input,
+                                              input.size(),
+                                              offset,
+                                              scalar.validity_data()};
+
+        thrust::transform_if(rmm::exec_policy(stream)->on(stream),
```

Comment:
It seems like this could be either `transform_if` or `transform` since the result will have the null mask applied either way. Which is faster?

Reply:
Are you saying we could use one or the other, and not both? If so, `transform` has the possibility of doing unnecessary writes. However, I don't know how well `transform__if` will get optimized if given it a predicate returning a constant.

Reply:
Right, I'm not sure if `transform_if` will be faster because it can skip writes, or possibly slower because it adds branching

Reply:
benchmark results of 
`transform_if(device_vector<int>(size),  , negate, (x%5==0 || x%7==0))` 
vs 
`transform(device_vector<int>(size), negate)`

```
Benchmark                                                      Time             CPU      Time Old      Time New       CPU Old       CPU New
-------------------------------------------------------------------------------------------------------------------------------------------
[BM_TransformIF vs. BM_Transform]/1024                      -0.0522         -0.0254        318273        301674        300298        292660
[BM_TransformIF vs. BM_Transform]/2048                      -0.0216         -0.0034        310464        303747        298596        297580
[BM_TransformIF vs. BM_Transform]/4096                      -0.0507         -0.0184        312541        296696        298178        292699
[BM_TransformIF vs. BM_Transform]/8192                      -0.0395         -0.0232        312513        300154        300359        293378
[BM_TransformIF vs. BM_Transform]/16384                     -0.0062         -0.0159        299061        297210        297275        292563
[BM_TransformIF vs. BM_Transform]/32768                     -0.0056         -0.0147        299699        298025        297382        293017
[BM_TransformIF vs. BM_Transform]/65536                     -0.0045         -0.0147        299246        297886        297664        293285
[BM_TransformIF vs. BM_Transform]/131072                    -0.0049         -0.0144        299494        298015        298242        293945
[BM_TransformIF vs. BM_Transform]/262144                    +0.0010         -0.0138        300863        301160        300026        295901
[BM_TransformIF vs. BM_Transform]/524288                    -0.0469         -0.0170        318365        303437        303903        298730
[BM_TransformIF vs. BM_Transform]/1048576                   -0.0092         +0.0007        318743        315811        311972        312189
[BM_TransformIF vs. BM_Transform]/2097152                   +0.0671         +0.0016        330638        352837        326486        327009
[BM_TransformIF vs. BM_Transform]/4194304                   -0.0554         -0.0058        388266        366763        363097        360995
[BM_TransformIF vs. BM_Transform]/8388608                   +0.0329         +0.0025        436858        451247        421882        422934
[BM_TransformIF vs. BM_Transform]/16777216                  +0.0089         +0.0015        574508        579632        540607        541406
[BM_TransformIF vs. BM_Transform]/33554432                  -0.0017         -0.0014        833935        832497        779302        778216
[BM_TransformIF vs. BM_Transform]/67108864                  -0.0210         -0.0008       1315215       1287579       1250784       1249793
[BM_TransformIF vs. BM_Transform]/134217728                 +0.0222         +0.0024       2219663       2269001       2194241       2199469
[BM_TransformIF vs. BM_Transform]/268435456                 +0.0086         +0.0014       4132712       4168415       4095391       4100975
[BM_TransformIF vs. BM_Transform]/536870912                 +0.0011         -0.0004       8020811       8029842       7967592       7964049
[BM_TransformIF vs. BM_Transform]/1073741824                +0.0023         -0.0002      15927640      15963848      15870558      15867546
[BM_TransformIF vs. BM_Transform]/2147483648                -0.0013         -0.0018      32595627      32552067      32515499      32455778
```

Reply:
@karthikeyann  How did you produce this output?

Reply:
I wrote a google benchmark.
```cpp
#include <benchmark/benchmark.h>
#include <thrust/transform.h>
#include <thrust/functional.h>
#include <thrust/execution_policy.h>
#include <thrust/sequence.h>
#include <thrust/device_vector.h>
#include "../benchmarks/synchronization/synchronization.hpp"

struct is_odd
{
  __host__ __device__
  bool operator()(int x)
  {
    return (x%5 == 0) || (x%7 == 0);
  }
};

static void BM_TransformIF(benchmark::State& state) {
  thrust::device_vector<int> D(state.range(0));
  thrust::negate<int> op;
  //thrust::identity<int> op;
  thrust::sequence(D.begin(), D.end());
  for (auto _ : state) {
    cuda_event_timer timer(state, true);
    thrust::transform_if(thrust::device, D.begin(), D.end(), D.begin(), op, is_odd()); // in-place transformation
  }
}
// Register the function as a benchmark
BENCHMARK(BM_TransformIF)->RangeMultiplier(2)->Range(1UL<<10, 1UL<<31);

// Define another benchmark
static void BM_Transform(benchmark::State& state) {
  thrust::device_vector<int> D(state.range(0));
  thrust::negate<int> op;
  //thrust::identity<int> op;
  thrust::sequence(D.begin(), D.end());

  for (auto _ : state) {
    cuda_event_timer timer(state, true);
    thrust::transform(thrust::device, D.begin(), D.end(), D.begin(), op);// in-place transformation
  }
}
BENCHMARK(BM_Transform)->RangeMultiplier(2)->Range(1UL<<10, 1UL<<31);

BENCHMARK_MAIN();
```
and ran
`googlebenchmark/tools/compare.py filter ./a.out BM_TransformIF BM_Transform`

Reply:
What's the conclusion, it doesn't matter very much which thrust transform function is used?

Reply:
BTW, every libcudf algorithm should really have a google benchmark.

Reply:
If it doesn't matter, the `transform_if` could be refactored with the `transform` above in the `scalar.is_valid() && not input.nullable()` branch, but it's not a big deal

Reply:
I'm going to benchmark this. I suspect the reason they perform similarly is that the predicate used in the example above is ineffective at preventing writes. The point of `transform_if` is to support the null fill value case, where for `shift(input, n, null_fills)` is guaranteed to have a run of `|n|` invalid elements.

Reply:
preliminary benchmarking suggests performance improves as `offset` increases. That's only with memory writes. However, shift is also reading from the input, so presumably that difference will get doubled.

Reply:
Benchmarking different implementations of `cudf::experimental::shift`, I've tried the following:
 - `transform only`
 - `transform + transform_if` (`transform_if` used when output is nullable)
 - `transform_if only`

Generally speaking:
- `transform + transform_if` is significantly faster for large inputs if the output will be nullable.
- `transform only` is faster for smaller data sets or if the output will be non-nullable.
- `transform_if only` is consistently slower, and has been excluded from the results.

```
`transform`:
----------------------------------------------------------------
shift_none/1024            0.027 ms        0.047 ms        25438
shift_none/4096            0.027 ms        0.047 ms        24795
shift_none/262144          0.040 ms        0.057 ms        17606
shift_none/16777216        0.639 ms        0.659 ms         1111
shift_none/1073741824       26.9 ms         27.0 ms           26
shift_half/1024            0.028 ms        0.047 ms        25519
shift_half/4096            0.028 ms        0.047 ms        25190
shift_half/262144          0.039 ms        0.057 ms        18143
shift_half/16777216        0.368 ms        0.388 ms         1903
shift_half/1073741824       21.3 ms         21.3 ms           33
shift_all/1024             0.027 ms        0.047 ms        25579
shift_all/4096             0.027 ms        0.047 ms        25644
shift_all/262144           0.036 ms        0.054 ms        19278
shift_all/16777216         0.267 ms        0.287 ms         2617
shift_all/1073741824        15.5 ms         15.5 ms           45
================================================================

`transform` + `transform_if`
----------------------------------------------------------------
shift_none/1024            0.032 ms        0.051 ms        22024
shift_none/4096            0.032 ms        0.051 ms        22053
shift_none/262144          0.047 ms        0.064 ms        14781
shift_none/16777216        0.629 ms        0.649 ms         1146
shift_none/1073741824       27.9 ms         27.9 ms           25
shift_half/1024            0.031 ms        0.051 ms        22178
shift_half/4096            0.031 ms        0.051 ms        22249
shift_half/262144          0.044 ms        0.062 ms        16035
shift_half/16777216        0.324 ms        0.344 ms         2159
shift_half/1073741824       19.0 ms         19.0 ms           37
shift_all/1024             0.031 ms        0.051 ms        22569
shift_all/4096             0.031 ms        0.051 ms        22761
shift_all/262144           0.039 ms        0.059 ms        18089
shift_all/16777216         0.189 ms        0.209 ms         3692
shift_all/1073741824        10.0 ms         10.0 ms           70
================================================================
```

Reply:
Given the performance improvements of `transform` in conjunction with `transform_if` for large nullable inputs, I think it is appropriate to go with that (current) implementation.

Reply:
Hypothetically there's a threshold where `|offset| > N` then `transform_if` is faster.

Reply:
Realizing the performance benefit here is _primarily_ due to `|offset|`, so I'm discarding `transform_if` in favor of using `transform` and adjusting begin/end iterators accordingly.